scripts/functions/similarity_metric.py

```diff
@@ -0,0 +1,662 @@
+Scripts to generate simulated data and visualize it
+'''
+
+import os
+import sys
+import ast
+import pandas as pd
+import numpy as np
+import random
+import glob
+import warnings
+warnings.filterwarnings(action='ignore')
+
+sys.path.append("../")
+from functions import cca_core
+
+from sklearn.decomposition import PCA
+from numpy.random import seed
+randomState = 123
+seed(randomState)
+
+
+def sim_svcca(simulated_data_file,
+              permuted_simulated_data_file,
+              num_experiments,
+              use_pca,
+              num_PCs,
+              base_dir,
+              analysis_name):
+    '''
+    We want to determine if the different batch simulated data is able to capture the
+    biological signal that is present in the original data:
+    How much of the real input data is captured in the simulated batch data?
+
+    In other words, we want to compare the representation of the real input data and the simulated batch data.
+    We will use **SVCCA** to compare these two representations.
+
+    Here, we apply Singular Vector Canonical Correlation Analysis
+    [Raghu et al. 2017](https://arxiv.org/pdf/1706.05806.pdf) [(github)](https://github.com/google/svcca)
+    to the UMAP and PCA representations of our batch 1 simulated dataset vs batch n simulated datasets.
+    The output of the SVCCA analysis is the SVCCA mean similarity score. This single number can be interpreted
+    as a measure of similarity between our original data vs batched dataset.
+
+    Briefly, SVCCA uses Singular Value Decomposition (SVD) to extract the components explaining 99% of the variation.
+    This is done to remove potential dimensions described by noise. Next, SVCCA performs a Canonical Correlation Analysis (CCA)
+     on the SVD matrices to identify maximum correlations of linear combinations of both input matrices.
+     The algorithm will identify the canonical correlations of highest magnitude across and within algorithms of the same dimensionality.
+
+    Arguments
+    ----------
+    simulated_data_file: str
+        File containing simulated gene expression data
+
+    permuted_simulated_data_file: str
+        File containing permuted simulated gene expression data
+
+    num_experiments: list
+        List of different numbers of experiments to add to
+        simulated data
+
+    use_pca: bool
+        True if want to represent expression data in top PCs before
+        calculating similarity
+
+    num_PCs: int
+        Number of top PCs to use to represent expression data
+
+    base_dir: str
+        Parent directory containing data files
+
+    analysis_name: str
+        Name of analysis. Format 'analysis_<int>'
+
+
+    Returns
+    --------
+    output_list: array
+        Similarity scores for each number of experiments added
+
+    permuted_svcca: float
+        Similarity score comparing the permuted data to the simulated data
+
+    '''
+
+    seed(randomState)
+
+    # Read in data
+    simulated_data = pd.read_table(
```

Comment:
lines 92 to 157 are identical to:
lines 259 - 324
lines 413 - 478
lines 568 - 633

pull it out as a separate function and call it four times instead. That'll cut ~ 240 lines to 4 lines and make it easier if you wanted to change it for whatever reason.